Test battery - new function for global config replacement.

hjoliver opened the pull request:
Adds a new function to `test_header`, for replacing the global config file in tests.  ~~It automatically adds existing `[hosts]` and `[suite host self-identification]` settings to the replacement file, because ignoring these items can break remote task tests (however, tests can still override these if necessary).~~

[UPDATE] It automatically adds settings in `global-tests.rc` to the new global config file, because some tests may require these in order to work (e.g. for some remote host related settings).

(Motivation: my desktop box disappeared from DNS at work, so I had to use `[suite host self-identification]` to self-identify to tasks by IP address, after which a bunch of tests still failed because they replaced `global.rc` and wiped out the new self-ident settings).

hjoliver commented:
(don'review just yet, I've got a couple of test failures...)

hjoliver commented on lib/cylc/task_proxy.py:
```diff
@@ -447,7 +447,7 @@ def _get_events_conf(self, key, default=None):
 
     def _get_host_conf(self, key, default=None):
         """Return a host setting from suite then global configuration."""
-        if self.tdef.rtconfig["remote"].get(key):
```
This could return the wrong result for boolean config items.

hjoliver commented on tests/events/11-cycle-task-event-job-logs-retrieve/suite.rc:
```diff
@@ -27,3 +27,6 @@ title=Task Event Job Log Retrieve
         inherit = T
+        [[[remote]]]
+            retrieve job logs = False
+
```
Makes this test independent of global config.

hjoliver commented:
All tests passing now.  @matthewrmshin - please assign reviews.

matthewrmshin commented:
These tests are failing for me. All appear to consistently time out.

```
./tests/cylc-message/00-ssh.t                             (Wstat: 0 Tests: 3 Failed: 1)
  Failed test:  2
./tests/events/10-task-event-job-logs-retrieve.t          (Wstat: 0 Tests: 5 Failed: 4)
  Failed tests:  2-5
./tests/events/11-cycle-task-event-job-logs-retrieve.t    (Wstat: 0 Tests: 4 Failed: 3)
  Failed tests:  2-4
./tests/cylc-kill/00-kill-multi-hosts.t                   (Wstat: 0 Tests: 3 Failed: 1)
  Failed test:  2
./tests/cylc-poll/04-poll-multi-hosts.t                   (Wstat: 0 Tests: 3 Failed: 1)
  Failed test:  2
./tests/cylc-cat-log/02-remote-custom-runtime-viewer-pbs.t (Wstat: 0 Tests: 2 Failed: 1)
  Failed test:  2
```

hjoliver commented:
I've just pushed up a quoting fix for `./tests/cylc-cat-log/02-remote-custom-runtime-viewer-pbs.t` (which I can't test without pbs) - env vars referenced inside single quotes instead of double.  The other tests are definitely passing here, on my VM at home and on the HPCF at work.  The difference vs master should be just retention of the two `global.rc` sections - so can you see if you have some site config setting that is disrupting the tests?

matthewrmshin commented:
`./tests/cylc-cat-log/02-remote-custom-runtime-viewer-pbs.t` is now good. I'll try and have a look to see why the others are failing.

hjoliver commented:
@matthewrmshin - I think perhaps a better way of handling the problem that I'm trying to address here would be to allow a second global conf file `global-test-battery.rc` that will be used by the test battery if it exists, otherwise my new function will generate a blank config file with any new bits appended.  That way your site shouldn't be affected at all by this change.  If you agree, I make this change.

matthewrmshin commented:
Sorry, I forgot to mention that these days I always run `cylc test-battery` twice, once as `CYLC_CONF_PATH= cylc test-battery` (so it does not pick up any global configuration), and once as `cylc test-battery` (so it does pick up the site and my global configurations).

Obviously, the tests only fail with our site global configurations. I'll take a look at it today to see what's affecting the failed tests.

matthewrmshin commented:
Strangely, the `retrieve job logs retry delays` settings are turned into minutes! We have something like this:

```
[hosts]
# ... snip ...
    [[myhost*]]
        retrieve job logs retry delays = PT10S, PT30S, PT3M
# ... snip ...
[test battery]
    remote host=myhost01
```

So, instead of an initial delay of 10 seconds, the initial delay has become 10.0 minutes. That's why all those suites were timing out.

matthewrmshin commented:
The tests all pass if I override the site global configuration with a user global configuration like this:

```
[hosts]
    [[myhost*]]
        retrieve job logs retry delays =
```

matthewrmshin commented:
Please feel free to try the `global-test-battery.rc` idea.

matthewrmshin commented:
`./tests/cylc-get-site-config/00-basic.t` test 2 is failing, with something like this:

```
Usage: cylc [admin] get-site-config [OPTIONS]

Print cylc site/user configuration settings.

By default all settings are printed. For specific sections or items
use -i/--item and wrap parent sections in square brackets:
   cylc get-site-config --item '[editors]terminal'
Multiple items can be specified at once.

cylc-get-site-config: error: ambiguous option: --print (--print-run-dir, --print-site-dir?)
```

hjoliver commented:
@matthewrmshin - this should be good now (at least it is here).  But, just push it back to 'soon' if you still have problems at your site.

matthewrmshin commented:
@benfitzpatrick do you mind doing a quick sanity check on this one?

benfitzpatrick commented on tests/cylc-cat-log/01-remote.t:
```diff
@@ -24,7 +24,10 @@ if [[ -z $CYLC_TEST_HOST ]]; then
 fi
 set_test_number 14
-export CYLC_CONF_PATH=
+create_test_globalrc '' $"
```
May as well be consistent by using `$'`

hjoliver replied:
Single quotes prevent shell variable evaluation in the string (so I should be consistent by using double double-quotes for both, but it looks a bit messier for the empty string case:`""`!)

benfitzpatrick commented on tests/lib/bash/test_header:
```diff
@@ -75,6 +75,12 @@
+#         Create a new global config file $PWD/conf from site global-tests.rc
+#         with PRE and POST pre- and ap-pended (PRE for top level items with no
+#         section heading). For multi-line strings in bash, use:
+#           $"the quick
```
The $ is not strictly necessary - only useful for backslash-expanding stuff?

hjoliver replied:
As far as I'm aware, it's necessary if you want to get a newline embedded in the string - am I wrong?

hjoliver replied:
OK, looks like I am wrong.  The `$"abc\ndef"` form is only needed to embed a newline as `\n`.

hjoliver replied:
I'll standardise these strings tomorrow...

matthewrmshin commented:
Tests now OK in my environment (with and without site/user configuration).

hjoliver commented:
@benfitzpatrick - I've addressed your comments, should be good to go now.  Tests still pass here.  Bumping back to 'soon', however, so I can get 6.8.1 out today.
